File: src/notification_tray/components/tray.py
import json
import logging
from datetime import UTC, datetime, timedelta
from pathlib import Path
from typing import Iterable

from PyQt5.QtCore import Qt
from PyQt5.QtGui import QColor, QFont, QIcon, QPainter, QPixmap
from PyQt5.QtWidgets import QAction, QApplication, QMenu, QSystemTrayIcon

from notification_tray.components.notification_cacher import NotificationCacher
from notification_tray.components.notifier import Notifier
from notification_tray.types.notification import NotificationFolder
from notification_tray.utils import settings
from notification_tray.utils.logging import log_input_and_output
from notification_tray.utils.settings import (get_notification_backoff_minutes,
                                              is_do_not_disturb_active,
                                              is_hide_from_tray_active)

from ..types.notification import CachedNotification

logger = logging.getLogger(__name__)


class Tray:
    def __init__(
        self,
        root_path: Path,
        do_not_disturb: settings.Cache,
        hide_from_tray: settings.Cache,
        notification_backoff_minutes: dict[Path, int],
        notifier: Notifier,
        notification_cacher: NotificationCacher,
        app: QApplication,
    ) -> None:
        self.do_not_disturb = do_not_disturb
        self.hide_from_tray = hide_from_tray
        self.notification_backoff_minutes = notification_backoff_minutes
        self.root_path = root_path
        self.tray_icon: QSystemTrayIcon | None = None
        self.tray_menu = QMenu()
        self.notifier = notifier
        self.app = app
        self.notification_cacher = notification_cacher
        logger.info(f"Started tray with root path {root_path}")

    def update_icon(self):
        logger.debug("Updating tray icon")

        def count_dir(dir_: NotificationFolder) -> int:
            logger.debug(f"Counting notifications in {dir_['path']}")
            if is_do_not_disturb_active(
                self.root_path, dir_["path"], self.do_not_disturb
            ) or is_hide_from_tray_active(
                self.root_path, dir_["path"], self.hide_from_tray
            ):
                logger.debug(f"DnD or hide from tray is active. Skipping")
                return 0
            else:
                notification_backoff_minutes = get_notification_backoff_minutes(
                    self.root_path, dir_["path"], self.notification_backoff_minutes
                )
                return sum(
                    1
                    for n in dir_["notifications"].values()
                    if not n.get("trashed")
                    and (
                        notification_backoff_minutes <= 0
                        or (datetime.now(UTC) - n["at"]).total_seconds() // 60
                        > notification_backoff_minutes
                    )
                ) + sum(
                    map(count_dir, dir_["folders"].values()),
                )

        file_count = count_dir(self.notification_cacher.notification_cache)
        if file_count == 0:
            if self.tray_icon is not None:
                self.tray_icon.hide()
        else:
            if self.tray_icon is None:
                self.tray_icon = QSystemTrayIcon(parent=self.app)
            self.tray_icon.setIcon(self.get_notification_badge(file_count))
            self.tray_icon.activated.connect(self.setup_tray_menu)
            self.tray_icon.show()
            self.tray_icon.setContextMenu(self.tray_menu)

    def get_notification_badge(self, number: int) -> QIcon:
        # Create a pixmap with a transparent background
        pixmap = QPixmap(40, 40)
        pixmap.fill(Qt.GlobalColor.transparent)

        # Create a painter
        painter = QPainter(pixmap)

        # Set up the font
        font = QFont()
        font.setPixelSize(24)
        font.setBold(True)
        painter.setFont(font)

        # Draw the red circle
        painter.setBrush(QColor(255, 0, 0))  # Red color
        painter.setPen(Qt.PenStyle.NoPen)
        painter.drawEllipse(0, 0, 40, 40)

        # Set the color for the text (white)
        painter.setPen(QColor(255, 255, 255))

        # Draw the number in the center of the circle
        painter.drawText(pixmap.rect(), Qt.AlignmentFlag.AlignCenter, str(number))

        # End painting
        painter.end()

        # Create and return the icon
        return QIcon(pixmap)

    def setup_tray_menu(self):
        self.tray_menu.clear()
        self.populate_submenu(self.tray_menu, self.notification_cacher.notification_cache)

        # Add exit option
        exit_action = QAction("Exit", self.tray_menu)
        exit_action.triggered.connect(self.app.quit)
        self.tray_menu.addAction(exit_action)  # type: ignore

    def add_directory_contents(self, path: Path, menu: QMenu):
        def has_notifications(dir_: NotificationFolder) -> bool:
            return (
                not is_do_not_disturb_active(
                    self.root_path, dir_["path"], self.do_not_disturb
                )
                and not is_hide_from_tray_active(
                    self.root_path, dir_["path"], self.hide_from_tray
                )
                and (
                    any(not n.get("trashed") for n in dir_["notifications"].values())
                    or (any(map(has_notifications, dir_["folders"].values())))
                )
            )

        notifications_cache = self.notification_cacher.notification_cache
        for folder in path.relative_to(self.root_path).parts:
            notifications_cache = notifications_cache["folders"][folder]
        for name, folder in sorted(notifications_cache["folders"].items()):
            if has_notifications(folder):
                submenu = QMenu(name, menu)
                menu.addMenu(submenu)
                placeholder = QAction("Loading...", submenu)
                submenu.addAction(placeholder)  # type: ignore
                submenu.aboutToShow.connect(
                    lambda sub=submenu, f=folder: self.populate_submenu(sub, f)  # type: ignore
                )
        for name, notification in notifications_cache["notifications"].items():
            if not notification.get("trashed"):
                file_action = QAction(name, menu)
                file_action.triggered.connect(
                    lambda checked, p=notification: self.notifier.notify(p, is_batch=True)  # type: ignore
                )
                menu.addAction(file_action)  # type: ignore

    def populate_submenu(self, submenu: QMenu, folder: NotificationFolder):
        path = folder["path"]
        if not (actions := submenu.actions()) or actions[0].text() == "Loading...":
            submenu.clear()
            self.add_directory_contents(path, submenu)

            # Re-add the "Move to Trash" action after populating
            trash_action = QAction("Move to Trash", submenu)
            trash_action.triggered.connect(
                lambda checked, p=path: self.notification_cacher.trash(p)  # type: ignore
            )
            submenu.addAction(trash_action)  # type: ignore

            show_all_action = QAction("Show All", submenu)
            show_all_action.triggered.connect(
                lambda checked, f=folder: self.notify(f)  # type: ignore
            )
            submenu.addAction(show_all_action)  # type: ignore

            # Add "Do Not Disturb" submenu
            dnd_submenu = QMenu("Do Not Disturb", submenu)
            submenu.addMenu(dnd_submenu)
            for text, duration in [
                ("1 hour", datetime.now(UTC) + timedelta(hours=1)),
                ("8 hours", datetime.now(UTC) + timedelta(hours=8)),
                ("Forever", datetime(9999, 1, 1, tzinfo=UTC)),
            ]:
                dnd_action = QAction(text, dnd_submenu)
                dnd_action.triggered.connect(
                    lambda checked, p=str(path), d=duration: self.update_datetime_setting(  # type: ignore
                        "do_not_disturb", p, d, cache=self.do_not_disturb
                    )  # type: ignore
                )
                dnd_submenu.addAction(dnd_action)  # type: ignore

            # Add "Hide From Tray" submenu
            dnd_submenu = QMenu("Hide From Tray", submenu)
            submenu.addMenu(dnd_submenu)
            for text, duration in [
                ("1 hour", datetime.now(UTC) + timedelta(hours=1)),
                ("8 hours", datetime.now(UTC) + timedelta(hours=8)),
                ("Forever", datetime(9999, 1, 1, tzinfo=UTC)),
            ]:
                dnd_action = QAction(text, dnd_submenu)
                dnd_action.triggered.connect(
                    lambda checked, p=str(path), d=duration: self.update_datetime_setting(  # type: ignore
                        "hide_from_tray_until", p, d, cache=self.hide_from_tray
                    )  # type: ignore
                )
                dnd_submenu.addAction(dnd_action)  # type: ignore

            # Add "Batch notifications" submenu
            dnd_submenu = QMenu("Batch Notifications", submenu)
            submenu.addMenu(dnd_submenu)
            for text, minutes in [
                ("Every minute", 1),
                ("Every 5 minutes", 5),
                ("Every 10 minutes", 10),
            ]:
                dnd_action = QAction(text, dnd_submenu)
                dnd_action.triggered.connect(
                    lambda checked, p=path, m=minutes: self.update_notification_backoff_minutes(  # type: ignore
                        p, m
                    )  # type: ignore
                )
                dnd_submenu.addAction(dnd_action)  # type: ignore

    def notify(self, folder: NotificationFolder):
        def get_notifications(
            folder: NotificationFolder,
        ) -> Iterable[CachedNotification]:
            yield from folder["notifications"].values()
            for folder in folder["folders"].values():
                yield from get_notifications(folder)

        logger.info(f"Displaying all notifications for folder {folder['path']}")
        self.notifier.notify(*get_notifications(folder), is_batch=True)

    @log_input_and_output(logging.DEBUG)
    def update_notification_backoff_minutes(self, folder_path: Path, minutes: int):
        folder_path_ = Path(folder_path).absolute()
        settings_file = folder_path_ / ".settings.json"
        self.notification_backoff_minutes[folder_path_] = minutes
        try:
            existing_settings = json.loads(settings_file.read_text())
        except FileNotFoundError:
            existing_settings = {}
        settings_file.write_text(
            json.dumps(existing_settings | {"notification_backoff_minutes": minutes})
        )

    @log_input_and_output(logging.DEBUG)
    def update_datetime_setting(
        self,
        setting_name: str,
        folder_path: str,
        until: datetime,
        cache: settings.Cache,
    ):
        settings.write_datetime_setting(folder_path, setting_name, until, cache=cache)
        settings.cache_datetime_setting(Path(folder_path), setting_name, cache=cache)
        self.update_icon()
        self.setup_tray_menu()

    def refresh(self) -> None:
        self.update_icon()
        self.setup_tray_menu()

File: src/tray.cpp
#include "tray.h"

#include "utils/logging.h"
#include "utils/settings.h"

#include <QAction>
#include <QColor>
#include <QFile>
#include <QFont>
#include <QJsonDocument>
#include <QJsonObject>
#include <QMenu>
#include <QPainter>

static Logger logger = Logger::getLogger("Tray");

Tray::Tray(const fs::path& root_path, const Cache& do_not_disturb, const Cache& hide_from_tray,
           const std::map<fs::path, int>& notification_backoff_minutes, Notifier* notifier,
           NotificationCacher* notification_cacher, QApplication* app, QObject* parent)
    : QObject(parent)
    , m_root_path(root_path)
    , m_do_not_disturb(do_not_disturb)
    , m_hide_from_tray(hide_from_tray)
    , m_notification_backoff_minutes(notification_backoff_minutes)
    , m_notifier(notifier)
    , m_notification_cacher(notification_cacher)
    , m_app(app) {
    m_tray_menu = new QMenu();
    logger.info(
        QString("Started tray with root path %1").arg(QString::fromStdString(root_path.string())));
}

int Tray::countDir(NotificationFolder& dir) {
    if (Settings::isDoNotDisturbActive(m_root_path, dir.path, m_do_not_disturb) ||
        Settings::isHideFromTrayActive(m_root_path, dir.path, m_hide_from_tray)) {
        return 0;
    }

    int count = 0;
    int backoff = Settings::getNotificationBackoffMinutes(m_root_path, dir.path,
                                                          m_notification_backoff_minutes);

    for (const auto& [name, notif] : dir.notifications) {
        if (notif.trashed)
            continue;

        if (backoff <= 0) {
            count++;
        } else {
            qint64 minutes_since = notif.at.secsTo(QDateTime::currentDateTimeUtc()) / 60;
            if (minutes_since > backoff) {
                count++;
            }
        }
    }

    for (auto& [name, folder] : dir.folders) {
        count += countDir(folder);
    }

    return count;
}

bool Tray::hasNotifications(NotificationFolder& dir) {
    if (Settings::isDoNotDisturbActive(m_root_path, dir.path, m_do_not_disturb) ||
        Settings::isHideFromTrayActive(m_root_path, dir.path, m_hide_from_tray)) {
        return false;
    }

    for (const auto& [name, notif] : dir.notifications) {
        if (!notif.trashed) {
            return true;
        }
    }

    for (auto& [name, folder] : dir.folders) {
        if (hasNotifications(folder)) {
            return true;
        }
    }

    return false;
}

void Tray::updateIcon() {
    logger.debug("Updating tray icon");

    int file_count = countDir(m_notification_cacher->notification_cache);
    logger.info(QString("Tray icon update: %1 notifications").arg(file_count));

    if (file_count == 0) {
        if (m_tray_icon) {
            m_tray_icon->hide();
        }
    } else {
        if (!m_tray_icon) {
            m_tray_icon = new QSystemTrayIcon(m_app);
        }
        m_tray_icon->setIcon(getNotificationBadge(file_count));
        connect(m_tray_icon, &QSystemTrayIcon::activated, this, &Tray::setupTrayMenu);
        m_tray_icon->show();
        m_tray_icon->setContextMenu(m_tray_menu);
    }
}

QIcon Tray::getNotificationBadge(int number) {
    QPixmap pixmap(40, 40);
    pixmap.fill(Qt::transparent);

    QPainter painter(&pixmap);

    QFont font;
    font.setPixelSize(24);
    font.setBold(true);
    painter.setFont(font);

    painter.setBrush(QColor(255, 0, 0));
    painter.setPen(Qt::NoPen);
    painter.drawEllipse(0, 0, 40, 40);

    painter.setPen(QColor(255, 255, 255));
    painter.drawText(pixmap.rect(), Qt::AlignCenter, QString::number(number));

    painter.end();

    return QIcon(pixmap);
}

void Tray::setupTrayMenu() {
    m_tray_menu->clear();
    populateSubmenu(m_tray_menu, m_notification_cacher->notification_cache);

    QAction* exit_action = new QAction("Exit", m_tray_menu);
    connect(exit_action, &QAction::triggered, m_app, &QApplication::quit);
    m_tray_menu->addAction(exit_action);
}

void Tray::addDirectoryContents(const fs::path& path, QMenu* menu) {
    fs::path relative = fs::relative(path, m_root_path);
    NotificationFolder* current = &m_notification_cacher->notification_cache;

    for (const auto& part : relative) {
        QString folder_name = QString::fromStdString(part.string());
        if (current->folders.count(folder_name)) {
            current = &current->folders[folder_name];
        }
    }

    for (auto& [name, folder] : current->folders) {
        // Skip folders with no visible notifications
        if (!hasNotifications(folder)) {
            continue;
        }

        QMenu* submenu = new QMenu(name, menu);
        menu->addMenu(submenu);

        QAction* placeholder = new QAction("Loading...", submenu);
        submenu->addAction(placeholder);

        // Store the folder path to avoid dangling reference
        fs::path folder_path = folder.path;
        connect(submenu, &QMenu::aboutToShow, [this, submenu, folder_path]() {
            // Only populate if not already done (check for placeholder)
            QList<QAction*> actions = submenu->actions();
            if (actions.isEmpty() || actions[0]->text() != "Loading...") {
                return;  // Already populated
            }

            // Find the folder in the cache by path
            NotificationFolder* current_folder = &m_notification_cacher->notification_cache;
            fs::path relative = fs::relative(folder_path, m_root_path);
            for (const auto& part : relative) {
                QString folder_name = QString::fromStdString(part.string());
                if (current_folder->folders.count(folder_name)) {
                    current_folder = &current_folder->folders[folder_name];
                } else {
                    return;  // Folder no longer exists
                }
            }

            populateSubmenu(submenu, *current_folder);
        });
    }

    for (auto& [name, notif] : current->notifications) {
        if (!notif.trashed) {
            QAction* file_action = new QAction(name, menu);
            connect(file_action, &QAction::triggered,
                    [this, notif]() { m_notifier->notify(notif, true); });
            menu->addAction(file_action);
        }
    }
}

void Tray::populateSubmenu(QMenu* submenu, NotificationFolder& folder) {
    // Only populate if showing "Loading..." placeholder (matches Python behavior)
    QList<QAction*> actions = submenu->actions();
    if (!actions.isEmpty() && actions[0]->text() != "Loading...") {
        return;  // Already populated
    }

    submenu->clear();
    addDirectoryContents(folder.path, submenu);

    fs::path folder_path = folder.path;  // Capture by value to avoid dangling reference

    QAction* trash_action = new QAction("Move to Trash", submenu);
    connect(trash_action, &QAction::triggered,
            [this, folder_path]() { m_notification_cacher->trash(folder_path); });
    submenu->addAction(trash_action);

    QAction* show_all = new QAction("Show All", submenu);
    connect(show_all, &QAction::triggered, [this, folder_path]() {
        // Find the folder in the cache by path
        NotificationFolder* current_folder = &m_notification_cacher->notification_cache;
        fs::path relative = fs::relative(folder_path, m_root_path);
        for (const auto& part : relative) {
            QString folder_name = QString::fromStdString(part.string());
            if (current_folder->folders.count(folder_name)) {
                current_folder = &current_folder->folders[folder_name];
            } else {
                return;  // Folder no longer exists
            }
        }
        notifyFolder(*current_folder);
    });
    submenu->addAction(show_all);

    // Add "Do Not Disturb" submenu
    QMenu* dnd_menu = new QMenu("Do Not Disturb", submenu);
    submenu->addMenu(dnd_menu);

    QDateTime now = QDateTime::currentDateTimeUtc();
    QList<QPair<QString, QDateTime>> dnd_options = {
        {"1 hour", now.addSecs(3600)},
        {"8 hours", now.addSecs(28800)},
        {"Forever", QDateTime(QDate(9999, 1, 1), QTime(), Qt::UTC)}};

    for (const auto& [text, time] : dnd_options) {
        QAction* action = new QAction(text, dnd_menu);
        connect(action, &QAction::triggered, [this, folder_path, time]() {
            updateDateTimeSetting("do_not_disturb_until", folder_path, time, m_do_not_disturb);
        });
        dnd_menu->addAction(action);
    }

    // Add "Hide From Tray" submenu
    QMenu* hide_menu = new QMenu("Hide From Tray", submenu);
    submenu->addMenu(hide_menu);

    for (const auto& [text, time] : dnd_options) {
        QAction* action = new QAction(text, hide_menu);
        connect(action, &QAction::triggered, [this, folder_path, time]() {
            updateDateTimeSetting("hide_from_tray_until", folder_path, time, m_hide_from_tray);
        });
        hide_menu->addAction(action);
    }

    // Add "Batch Notifications" submenu
    QMenu* batch_menu = new QMenu("Batch Notifications", submenu);
    submenu->addMenu(batch_menu);

    QList<QPair<QString, int>> batch_options = {
        {"Every minute", 1}, {"Every 5 minutes", 5}, {"Every 10 minutes", 10}};

    for (const auto& [text, minutes] : batch_options) {
        QAction* action = new QAction(text, batch_menu);
        connect(action, &QAction::triggered, [this, folder_path, minutes]() {
            updateNotificationBackoffMinutes(folder_path, minutes);
        });
        batch_menu->addAction(action);
    }
}

void Tray::notifyFolder(NotificationFolder& folder) {
    std::vector<CachedNotification> notifications;

    std::function<void(NotificationFolder&)> collect = [&](NotificationFolder& f) {
        for (auto& [name, notif] : f.notifications) {
            notifications.push_back(notif);
        }
        for (auto& [name, subfolder] : f.folders) {
            collect(subfolder);
        }
    };

    collect(folder);
    m_notifier->notify(notifications, true);
}

void Tray::updateNotificationBackoffMinutes(const fs::path& folder_path, int minutes) {
    m_notification_backoff_minutes[folder_path] = minutes;

    fs::path settings_file = folder_path / ".settings.json";
    QString settings_path = QString::fromStdString(settings_file.string());

    QJsonObject existing_settings;
    QFile file(settings_path);
    if (file.open(QIODevice::ReadOnly)) {
        existing_settings = QJsonDocument::fromJson(file.readAll()).object();
        file.close();
    }

    existing_settings["notification_backoff_minutes"] = minutes;

    if (file.open(QIODevice::WriteOnly)) {
        file.write(QJsonDocument(existing_settings).toJson());
        file.close();
    }
}

void Tray::updateDateTimeSetting(const QString& setting_name, const fs::path& folder_path,
                                 const QDateTime& until, Cache& cache) {
    Settings::writeDateTimeSetting(folder_path, setting_name, until, cache);
    Settings::cacheDateTimeSetting(folder_path, setting_name, cache);
    updateIcon();
    setupTrayMenu();
}

void Tray::refresh() {
    logger.info("Tray refresh called");
    updateIcon();
    setupTrayMenu();
}
